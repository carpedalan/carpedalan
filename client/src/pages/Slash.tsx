import Feed from 'components/Feed';
import Grid from 'components/Grid';
import debug from 'debug';
import usePosts, { PostsWithTagsWithFakes } from 'hooks/usePosts';
import useRouter from 'hooks/useRouter';
import * as React from 'react';
import { Link, Route } from 'react-router-dom';
import { default as styled } from 'styled-components';
import Gallery from 'components/Gallery';
const log = debug('component:Slash');

const { useState, useLayoutEffect, lazy, Fragment } = React;

// const LazyGallery = lazy(() =>
//   import(/* webpackChunkName: "gallery" */ 'components/Gallery'),
// );

const InnerWrapper = styled.main`
  max-width: 768px;
  margin: auto;
  height: 100%;
`;

const { useEffect, useRef } = React;

const Wrapper = styled.div`
  width: 100%;
  height: 100%;
`;

const PostWrapper = styled.article`
  max-width: 620px;
  width: 100%;
  margin: 0 auto;
`;

const GridListSwitcher = styled.div`
  position: fixed;
  z-index: 2;
  top: 0;
  right: 0;
`;

const RowWrapper = styled.div`
  display: flex;
`;

interface RowRender {
  index: number;
  style: React.CSSProperties;
}

const Slash: React.FC = (): React.ReactElement => {
  const { request, posts, loading, response } = usePosts();
  const [postsWithTitle, setPostsWithTitle] = useState<
    PostsWithTagsWithFakes[]
  >(posts);

  const {
    location: { hash, pathname },
  } = useRouter();

  function isGrid() {
    return hash.includes('grid');
  }

  useEffect(() => {
    log('%c post dep changed', 'background: blue;');
    const newPosts = [...posts];
    newPosts.unshift({ fake: true, placeholder: '', key: 'title' });
    setPostsWithTitle(newPosts);
  }, [posts]);

  return (
    <Fragment>
      <GridListSwitcher>
        <Link to={`${pathname}${hash.includes('grid') ? '' : '#grid'}`}>
          {hash.includes('grid') ? 'List' : 'Grid'}
        </Link>
      </GridListSwitcher>

      <Wrapper data-testid="home">
        {isGrid() ? (
          <Grid itemsWithTitle={postsWithTitle} />
        ) : (
          <Feed itemsWithTitle={postsWithTitle} />
        )}
      </Wrapper>
      <Route exact={true} path="**/gallery/:postId" component={Gallery} />
    </Fragment>
  );
};

export default Slash;
