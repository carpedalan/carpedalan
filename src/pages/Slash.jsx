import React, { useContext, useEffect, useRef, useState } from 'react';
import {
  AutoSizer,
  InfiniteLoader,
  List,
  WindowScroller,
} from 'react-virtualized';
import throttle from 'lodash/throttle';

import Wrapper from '../styles/Wrapper';
import { MEDIUM } from '../../shared/constants';
import { Posts } from '../providers/PostsProvider';
import { Tag } from '../providers/TagProvider';
import PostRenderer from '../components/PostRenderer';
import Title from '../styles/Title';
import log from '../utils/log';

let lastValue = 0;
const throttled = throttle((e, setShouldShowImages) => {
  const delta = Math.abs(e.scrollTop - lastValue);
  lastValue = e.scrollTop;
  setShouldShowImages(e.scrollTop < 100 || delta < 2500);
}, 250);

export default function Slash() {
  const { getPosts, posts, meta, cache, patchPost, delPost } = useContext(
    Posts,
  );
  const { tags } = useContext(Tag);
  // const { user } = useContext(User);
  // const [query, setQuery] = useState(null);
  const [isEditing, setEditing] = useState(false);
  const [shouldShowImages, setShouldShowImages] = useState(true);
  const listRef = useRef(null);
  const throttledResize = throttle(() => {
    if (listRef.current) {
      cache.clearAll();
    }
  }, 250);

  window.addEventListener('resize', throttledResize);
  useEffect(() => {
    getPosts(1);
    return () => window.removeEventListener('resize', throttledResize);
  }, []);

  useEffect(
    () => {
      cache.clearAll();
      listRef.current.recomputeRowHeights(0);
    },
    [isEditing],
  );

  const handlePatchPost = id => async values => {
    await patchPost(id)(values);
    try {
      listRef.current.forceUpdate();
      setEditing(false);
    } catch (e) {
      log.error(e);
    }
  };

  // const delPost = id => async () => {
  //   await request.delete(`${API_PATH}/posts/${id}`);
  //   setCacheValid(false);
  //   await getPosts(1);

  //   cache.clearAll();
  //   listRef.forceUpdateGrid();
  // };

  // const getProps = (description, id) =>
  //   isEditing
  //     ? { initial: { [DESCRIPTION]: description }, onSubmit: handleSubmit(id) }
  //     : {};

  function isRowLoaded({ index }) {
    return !!posts[index];
  }

  function loadMoreRows() {
    return getPosts();
  }

  const handleScroll = onChildScroll => e => {
    onChildScroll(e);
    throttled(e, setShouldShowImages);
  };
  return (
    <Wrapper>
      <Title center size="large">
        Carpe Dalan
      </Title>
      <WindowScroller>
        {({ height, isScrolling, registerChild, onChildScroll, scrollTop }) => (
          <AutoSizer disableHeight>
            {({ width }) => (
              <InfiniteLoader
                isRowLoaded={isRowLoaded}
                loadMoreRows={loadMoreRows}
                rowCount={meta.count}
              >
                {({
                  onRowsRendered /* , registerChild: registerInfiniteChild */,
                }) => (
                  <div ref={registerChild}>
                    <List
                      width={width}
                      height={height}
                      autoHeight
                      ref={listRef}
                      onRowsRendered={onRowsRendered}
                      deferredMeasurementCache={cache}
                      onScroll={handleScroll(onChildScroll)}
                      rowHeight={cache.rowHeight}
                      rowRenderer={PostRenderer}
                      rowCount={meta.count}
                      overscanRowCount={10}
                      isScrolling={isScrolling}
                      scrollTop={scrollTop}
                      posts={posts}
                      cache={cache}
                      shouldShowImages={shouldShowImages}
                      size={MEDIUM}
                      showDescription
                      isAdmin
                      setEditing={setEditing}
                      isEditing={isEditing}
                      delPost={delPost}
                      patchPost={handlePatchPost}
                      tags={tags}
                    />
                  </div>
                )}
              </InfiniteLoader>
            )}
          </AutoSizer>
        )}
      </WindowScroller>
    </Wrapper>
  );
}
