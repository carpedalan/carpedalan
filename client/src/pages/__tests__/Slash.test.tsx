import axios, { AxiosPromise, AxiosResponse, AxiosStatic } from 'axios';
import * as React from 'react';
import { MemoryRouter, Redirect, Route } from 'react-router';
import { BrowserRouter, Link } from 'react-router-dom';
import {
  act,
  cleanup,
  fireEvent,
  render,
  waitForElement,
} from '@testing-library/react';

import Slash from '../Slash';

const mockedAxios = axios as jest.Mocked<typeof axios>;

jest.mock('axios');

describe('<Slash />', () => {
  beforeEach(jest.resetAllMocks);
  it('should match snapshot', () => {
    const app = render(
      <MemoryRouter>
        <Slash />
      </MemoryRouter>,
    );
    expect(app).toMatchSnapshot();
  });
  // it('should')
});
