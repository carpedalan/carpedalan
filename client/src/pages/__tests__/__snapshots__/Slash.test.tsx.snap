// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slash /> should match snapshot 1`] = `
<Fragment>
  <styled.div>
    <Link
      to="#grid"
    >
      Grid
    </Link>
  </styled.div>
  <styled.div
    data-testid="home"
  >
    <Feed
      itemsWithTitle={Array []}
    />
  </styled.div>
  <Route
    component={[Function]}
    exact={true}
    path="**/gallery/:postId"
  />
</Fragment>
`;
